server/app/services/applicant/ApplicantService.java

```diff
@@ -629,32 +646,110 @@ private ApplicationPrograms relevantProgramsForApplicant(
+                  programNamesWithApplicationsFuture.toArray(
+                      new CompletableFuture[programNamesWithApplicationsFuture.size()]))
+              .thenAcceptAsync(
+                  (v) -> {
+                    Set<String> unappliedActivePrograms =
+                        Sets.difference(activeProgramNames.keySet(), programNamesWithApplications);
+                    Long appId = applications.stream().findFirst().get().getApplicant().id;
+
+                    ApplicationPrograms.Builder appProgramsBuilder =
+                        ApplicationPrograms.builder()
+                            .setInProgress(sortByProgramId(inProgressPrograms.build()))
+                            .setSubmitted(sortByProgramId(submittedPrograms.build()));
+
+                    if (unappliedActivePrograms.isEmpty()) {
+                      applicationProgramsFuture.complete(
+                          appProgramsBuilder
+                              .setUnapplied(sortByProgramId(unappliedPrograms.build()))
+                              .build());
+                    }
+
+                    ArrayList<CompletableFuture<String>> unappliedProgramsFuture =
+                        new ArrayList<>();
+                    unappliedActivePrograms.forEach(
+                        programName -> {
+                          CompletableFuture<String> programNameFuture = new CompletableFuture<>();
+                          ProgramDefinition program = activeProgramNames.get(programName);
+                          ApplicantProgramData.Builder applicantProgramDataBuilder =
+                              ApplicantProgramData.builder().setProgram(program);
+                          unappliedProgramsFuture.add(programNameFuture);
+                          setApplicationsForType(
+                                  appId, program, unappliedPrograms, applicantProgramDataBuilder)
+                              .thenApplyAsync((a) -> programNameFuture.complete(programName))
+                              .exceptionally(
+                                  ex -> {
+                                    throw new RuntimeException(ex);
+                                  });
+                        });
+                    var unusedUnappliedFuture =
+                        CompletableFuture.allOf(
+                                unappliedProgramsFuture.toArray(
+                                    new CompletableFuture[unappliedProgramsFuture.size()]))
+                            .thenAcceptAsync(
+                                (b) -> {
+                                  applicationProgramsFuture.complete(
+                                      appProgramsBuilder
+                                          .setUnapplied(sortByProgramId(unappliedPrograms.build()))
+                                          .build());
+                                });
+                  });
+    }
+    try {
+      return applicationProgramsFuture.get();
```

Comment:
**issue:** here and a couple other places there are calls to `CompleteableFuture.get` which will block the current thread.

The code here is pretty complex, making it all the harder. This page is probably the biggest performance risk in the system.

I wonder if it's feasible to do more of the DB querying in the beginning and then collate the results synchronously without having to work with futures.

Reply:
SO much better getting the data upfront - thanks for the suggestion. I didn't realize there were synchronous versions of the getReadOnlyApplicantProgramService function, but using those made it so that as long as I had the question data loaded in the program definitions, I could still initiate that class and use its methods.

Reply:
Great work cleaning this up, I agree it's much improved!